src/lib/Builder.hs

```diff
@@ -425,27 +425,28 @@ buildBlock
   :: ScopableBuilder m
   => (forall l. (Emits l, DExt n l) => m l (Atom l))
   -> m n (Block n)
-buildBlock cont = do
-  Abs decls results <- buildScoped do
-    result <- cont
-    ty <- {-# SCC blockTypeNormalization #-} cheapNormalize =<< getType result
-    return $ result `PairE` ty
-  let (result `PairE` ty) = results
-  let msg = "Decls:" <+> line <> pretty decls <> line
-            <> "Result:" <+> line <> pretty result <> line
-            <> "Of type:" <+> line <> pretty ty
-  ty' <- liftHoistExcept' (docAsStr msg) $ hoist decls ty
-  return $ Block (BlockAnn ty') decls result
-
-makeBlock :: EnvReader m => Nest Decl n l -> Atom l -> m l (Block n)
-makeBlock decls atom = do
-  ty <- {-# SCC blockTypeNormalization #-} cheapNormalize =<< getType atom
-  return $ makeBlockOfType decls atom ty
+buildBlock cont = buildScoped (cont >>= withType) >>= computeAbsEffects >>= absToBlock
+
+withType :: (EnvReader m, HasType e) => e l -> m l ((e `PairE` Type) l)
+withType e = do
+  ty <- {-# SCC blockTypeNormalization #-} cheapNormalize =<< getType e
+  return $ e `PairE` ty
+{-# INLINE withType #-}
+
+makeBlock :: Nest Decl n l -> EffectRow l -> Atom l -> Type l -> Block n
+makeBlock decls effs atom ty = Block (BlockAnn ty' effs') decls atom where
+  ty' = ignoreHoistFailure $ hoist decls ty
+  effs' = ignoreHoistFailure $ hoist decls effs
 {-# INLINE makeBlock #-}
 
-makeBlockOfType :: Nest Decl n l -> Atom l -> Type l -> Block n
-makeBlockOfType decls atom ty = Block (BlockAnn ty') decls atom
-  where ty' = ignoreHoistFailure $ hoist decls ty
+absToBlock :: (Fallible m) => Abs (Nest Decl) (EffectRow `PairE` (Atom `PairE` Type)) n -> m (Block n)
+absToBlock (Abs decls (effs `PairE` (result `PairE` ty))) = do
+  let msg = "Block:" <> nest 1 (prettyBlock decls result) <> line
+            <> group ("Of type:" <> nest 2 (line <> pretty ty)) <> line
+            <> group ("With effects:" <> nest 2 (line <> pretty effs))
+  ty' <- liftHoistExcept' (docAsStr msg) $ hoist decls ty
+  effs' <- liftHoistExcept' (docAsStr msg) $ hoist decls effs
+  return $ Block (BlockAnn ty' effs') decls result
```

Comment:
Can you please at least mark this as `INLINABLE`? It still runs in some arbitrary monad and we should at least open up the opportunities to specialization, though I imagine this will be inhibited by the huge `msg` expression in its body

Reply:
Done.